core/src/server/rpc_module.rs

```diff
@@ -851,25 +832,69 @@ pub struct SubscriptionSink {
 	close_notify: Option<SubscriptionPermit>,
 	/// MethodCallback.
 	method: &'static str,
-	/// Unique subscription.
-	uniq_sub: SubscriptionKey,
 	/// Shared Mutex of subscriptions for this method.
 	subscribers: Subscribers,
-	/// Future that returns when the unsubscribe method has been called.
-	unsubscribe: watch::Receiver<()>,
+	/// Unique subscription.
+	uniq_sub: SubscriptionKey,
+	/// The state of the subscription sink.
+	state: SubscriptionSinkState,
 	/// Claimed resources.
 	_claimed: Option<ResourceGuard>,
 }
 
 impl SubscriptionSink {
-	/// Send a message back to subscribers.
+	/// Reject the subscription call from [`ErrorObject`].
+	pub fn reject(&mut self, err: impl Into<ErrorObjectOwned>) -> Result<(), SubscriptionEmptyError> {
+		let id = self.state.id()?;
+
+		if self.inner.send_error(id, err.into()) {
+			self.state = self.state.reject()?;
+			Ok(())
+		} else {
+			Err(SubscriptionEmptyError)
+		}
+	}
+
+	/// Attempt to accept the subscription and respond the subscription method call.
 	///
-	/// Returns `Ok(true)` if the message could be send
-	/// Returns `Ok(false)` if the sink was closed (either because the subscription was closed or the connection was terminated)
-	/// Return `Err(err)` if the message could not be serialized.
+	/// Fails if the connection was closed, or if called multiple times.
+	pub fn accept(&mut self) -> Result<(), SubscriptionEmptyError> {
```

Comment:
I'm not really sure what the value of having `accept()` and `maybe_accept()` is, especially given both have the same return type and the error is not informative; I think there is potential for confusion over when to use each.

Is it worth returning a more informative error here (all errors convert into `SubscriptionEmptyError` anyway, which I view as a bit of a hack to help ergonomics in the common case, and don't think we should be using ourselves unless we really do have no information to return)?

I guess I'd go for something like just having `accept(&mut self) -> Result<(), AcceptError>` where `AcceptError` can be something like:

```rust
enum AcceptError {
    AlreadyAccepted,
    RemotePeerAborted,
}
```

(all names and such up for grabs, and should this be nested in a larger error type?)

That way, people can just call it and ignore the error if they want, or pay attention to the error to know why it didn't work out, and no need for a separate `maybe_accept`?

WDYT?

Reply:
these errors are still not propagated because of `SubscriptionEmptyError` so the error themselves are just swallowed here.

Thus, I don't get why need to extra verbose stuff `SubscriptionSinkState` but yeah if we want to propagate the error `AcceptError` makes more sense.